erdpy/contracts.py

```diff
@@ -177,16 +180,33 @@ def _interpret_return_data(self, data: str) -> Any:
 
 
 def _prepare_argument(argument: Any):
-    as_string = str(argument).upper()
-
-    if as_string.startswith(HEX_PREFIX):
-        return _prepare_hexadecimal(as_string)
-
-    return _prepare_decimal(as_string)
+    as_str = str(argument)
+    as_hex = _to_hex(as_str)
+    return as_hex
+
+
+def _to_hex(arg: str):
```

Comment:
Although this is indeed a big improvement and helps a lot for common scenarios, the assumptions that "true" / "false" are boolean values (and never plain strings), or "erd1..." is always an address (and never a string) could lead to some issues in the future.

Currently, erdpy doesn't handle contract ABIs and cannot do informed coding / decoding of arguments (as erdjs does).

However, until then, how do you feel about changing your implementation as follows?

```
prepare_argument(argument):
    if argument starts with "bool:", do top level encoding of a boolean (as you did in your implementation)
    else if argument starts with "address:", then do top level encoding of an address (as you did in your implementation)
    else if argument starts with "ascii:", then ...
    else if argument starts with "utf8:", then ...
    else if argument starts with "bytes:", then ... (assert hexadecimal input here)
    else if argument starts with "int:", then ...
    else if argument starts with "uint:", then ...
    (fallback for backwards compatibility) else if argument starts with HEX_PREFIX (as before, old implementation)
    (fallback for backwards compatibility) else if argument is numeric ...
```

Then, the user would be able to do as follows:

```
--arguments bool:true ascii:foobar bytes:ABBA uint:7 0x0BAD 42
```

Of course, we'll then deprecate the old (current) way in upcoming versions.

CC: @claudiu725 @dorin-iancu - how does this sound in your opinion (until we do proper ABI handling)?

References:
 - https://docs.elrond.com/developers/developer-reference/elrond-serialization-format/

Reply:
I agree with adding prefixes - sometimes the arguments can be ambiguous, especially if we allow any kind of strings to be passed as arguments. However, if we require strings to have a prefix, then the other types can be implied safely.
The prefix for strings, which we already use in mandos scenarios is `str:`. With this, there's no ambiguity between `true` and `str:true`.
The arguments above would look like:
`--arguments true str:foobar 0xABBA 7 0x0BAD 42`

Note: Later on, in another PR, we could add support for concatenation and fixed-width and nested types, bringing this closer to the existing mandos scenario syntax (eg. `u8:42|nested:str:hello|biguint:42`)

Reply:
I'm down to change the implementation @andreibancioiu. I thought the same as @claudiu725, adding the `str/ascii` prefix should be enough for this iteration.

Reply:
:rocket: 

Extra reference (from @andrei-marinica), which might help as well (now or in future PRs):

https://docs.elrond.com/developers/mandos-reference/values-simple/

Reply:
added the `str:` prefix.  saw that some tests failed. passing locally.

Reply:
can be re-reviewed imo. mandos test failing in erdpy-build workflow. re tested locally and passes, please have a look and let me know if it's something i could do or it's workflow related.
local versions: `rustc:1.60.0-nightly`, `vm tools: v1.4.39`

Reply:
The tests were failing because vm tools `v1.4.42` which was required by the templates' test was not released yet.
As it has been released today, the tests now pass.

It's ready for review.